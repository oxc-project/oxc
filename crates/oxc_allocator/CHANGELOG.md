# Changelog

All notable changes to this crate will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project does not adhere to [Semantic Versioning](https://semver.org/spec/v2.0.0.html) until v1.0.0.

## [0.6.0] - 2024-02-03

### Documentation

- Document behaviour of `Box`

## [0.5.0] - 2024-01-12

### Features

- Add `From` API (#1908)

## [0.4.0] - 2023-12-08

### Refactor

- Move to workspace lint table (#1444)

## [0.2.0] - 2023-09-14

### Refactor

- Improve code coverage a little bit
- Improve code coverage in various places (#721)

