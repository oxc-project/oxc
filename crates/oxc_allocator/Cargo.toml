[package]
name                   = "oxc_allocator"
version                = "0.11.0"
authors.workspace      = true
description.workspace  = true
edition.workspace      = true
homepage.workspace     = true
keywords.workspace     = true
license.workspace      = true
repository.workspace   = true
rust-version.workspace = true
categories.workspace   = true

[lints]
workspace = true

[lib]
doctest = false

[dependencies]
bumpalo = { workspace = true, features = ["collections"] }
serde   = { workspace = true }

[dev-dependencies]
serde_json = { workspace = true }
