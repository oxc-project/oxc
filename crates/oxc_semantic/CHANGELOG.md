# Changelog

All notable changes to this crate will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project does not adhere to [Semantic Versioning](https://semver.org/spec/v2.0.0.html) until v1.0.0.

## [0.11.1] - 2024-04-03

### Bug Fixes

- Flag function expressions with `SymbolFlags::Function` (#2891)

## [0.11.0] - 2024-03-30

### Features

- Add `Span` for JSDoc, JSDocTag (#2815)
- Distinguish type imports in ModuleRecord (#2785)

### Bug Fixes

- Ignore export declaration in no-duplicates (#2863)
- Missing SymbolFlags::Export when identifier used in ExportDefaultDeclaration (#2837)
- Incorrect ExportEntry span for ExportAllDeclaration in ModuleRecord (#2793)
- ModuleRecord's indirect_export_entires missing reexported imports (#2792)

### Refactor

- Distinguish whether requested_modules is type imports/exports (#2848)
- JSDocTag parser rework (#2765)

## [0.10.0] - 2024-03-14

### Features

- Merge features `serde` and `wasm` to `serialize` (#2716)
- Remove `From<String>` and `From<Cow>` API because they create memory leak (#2628)
- Move redeclare varaibles to symbol table (#2614)

### Bug Fixes

- Support multibyte chars (#2694)
- Fix up builder (#2623)

### Refactor

- Remove unused dependencies (#2718)
- Refactor `Trivias` API - have less noise around it (#2692)
- Reduce `cfg_attr` boilerplate with `SerAttrs` derive (#2669)
- Import `Tsify` to shorten code (#2665)
- "wasm" feature enable "serde" feature (#2639)
- Shorten manual TS defs (#2638)
- Make `CompactStr` immutable (#2620)
- Rename `CompactString` to `CompactStr` (#2619)

## [0.9.0] - 2024-03-05

### Features

- Remove all commonjs logic for import plugin (#2537)
- Call build module record (#2529)

### Bug Fixes

- Jsx reference with an incorrect node id (#2546)
- Incorrect scope for switch statement (#2513)

### Refactor

- Misc fixes for JSDoc related things (#2531)
- Replace InlinableString with CompactString for `Atom` (#2517)

## [0.8.0] - 2024-02-26

### Features

- Handle cjs `module.exports = {} as default export (#2493)
- Handle cjs `module.exports.foo = bar` and `exports.foo = bar` (#2492)
- Handle top-level `require` for import plugin (#2491)
- Add check for duplicate class elements in checker (#2455)
- Add static property, ElementKind::Getter, ElementKind::Setter in ClassTable (#2445)

### Bug Fixes

- Add export symbol flag to identifiers in export declarations (#2508)
- Improve import/no-named-as-default (#2494)
- Should return nearest JSDoc (#2490)
- Refactor jsdoc finding (#2437)
- Incorrect reference flag for MemberExpression assign (#2433)

### Performance

- Reduce visit parent nodes in resolve_reference_usages (#2419)

### Refactor

- S/NumberLiteral/NumericLiteral to align with estree
- S/ArrowExpression/ArrowFunctionExpression to align estree
- Remove `panic!` from examples (#2454)
- Delete the redundant code in binder (#2423)
- Reduce allocation in resolve_references_for_current_scope (#2414)
- Check directive by current_scope_id (#2411)

## [0.7.0] - 2024-02-09

### Features

- Add export binding for ExportDefaultDeclarations in module record (#2329)
- Enter AstKind::ExportDefaultDeclaration, AstKind::ExportNamedDeclaration and AstKind::ExportAllDeclaration (#2317)
- Report parameter related errors for setter/getter (#2316)
- Report type parameter list cannot be empty (#2315)
- Report unexpected type annotation in ArrayPattern (#2309)
- Apply ImportSpecifier's binder and remove ModuleDeclaration's binder (#2307)
- Fix memory leak by implementing inlineable string for oxc_allocator (#2294)

### Bug Fixes

- Remove unnecessary SymbolFlags::Import (#2311)
- Remove ignore cases (#2300)

## [0.6.0] - 2024-02-03

### Features

- Report no class name error (#2273)
- Check parameters property (#2264)
- Check optional parameters (#2263)
- Report error on optional variable declaration in TypeScript (#2261)
- Improve sample visualization (#2251)
- Track cfg index per ast node (#2210)
- Remove serde skip for symbol_id and reference_id (#2220)
- TypeScript definition for wasm target (#2158)
- Cfg prototype (#2019)
- Improve no_redeclare rule implementation (#2084)
- Remove import if only have type reference (#2001)

### Bug Fixes

- Handle short-circuiting operators in CFG (#2252)
- Proper traversal of try statements (#2250)
- Fix incorrect semantic example (#2198)
- Replace ClassStatickBlockAwait with ClassStaticBlockAwait (#2179)
- Print `Directive` original string (#2157)
- Incorrect reference flag (#2057)

### Refactor

- Adding binder for ImportSpecifier replaces the ModuleDeclaration's binder (#2230)
- Get function by scope_id in set_function_node_flag (#2208)
- Checking label in ContinueStatement based on LabelBuilder (#2202)
- Use LabelBuilder instead of UnusedLabeled (#2184)
- Move all miette usages to `oxc_diagnostics`
- Remove all #[dead_code[ from tester
- Rename RestElement to BindingRestElement (#2116)
- Add binder for FormalParameters and RestElement, replacing the binder for FormalParameters (#2114)
- Improve declare symbol logic in FormalParameters (#2088)

## [0.5.0] - 2024-01-12

### Features

- Support es2015 new target (#1967)
- Allow reserved keyword defined in ts module block (#1907)
- Add current_scope_flags function in SemanticBuilder (#1906)
- Visualize symbol (#1886)
- Visualize scope (#1882)
- Improve check super implementation, reduce access nodes (#1827)
- Support get node id by scope id (#1826)
- Support eslint/no-unused-private-class-members rule (#1820)
- Add ClassTable (#1793)
- Add SymbolFlags::Function for FunctionDeclaration (#1713)

### Bug Fixes

- Remove duplicate errors in ModuleDeclaration::ImportDeclaration (#1846)

### Performance

- Check duplicate parameters in Binder of FormalParameters (#1840)
- Just need to find the AstKind::FormalParameter in is_in_formal_parameters (#1852)
- Reduce calls to span() (#1851)
- Find class node by symbols in get_parent_es6_component (#1657)

### Refactor

- Improve ClassTable implmention and merge properties and methods to elements (#1902)
- Improve check function declaration implementation (#1854)
- Rename `add_node_id` to `add_current_node_id_to_current_scope` (#1847)
- Improve check private identifier implementation (#1794)
- Remove unused methods from `AstNode`

## [0.4.0] - 2023-12-08

### Features

- Support scope descendents starting from a certain scope. (#1629)
- Parse jsdoc on `PropertyDefinition` (#1517)
- Add the basics of comment printing (#1313)
- Add to_string function to VariableDelcartionKind (#1303)

### Refactor

- Move to workspace lint table (#1444)
- VariableDeclarationKind::to_string -> as_str (#1321)

## [0.3.0] - 2023-11-06

### Features

- Read comment pragma @jsxRuntime classic / automatic (#1133)
- Implement some of needs_explicit_esm for typescript (#1047)
- Bind function expression name (#1049)
- Add utils to make logical_assignment_operators pass (#1017)
- ES2020 Nullish Coalescing Operator (#1004)
- Finish 2016 exponentiation operator (#996)
- Check non-simple lhs expression of assignment expression (#994)
- Re-enable mangler (#972)
- Partially re-enable minifier (#963)
- Add no-redeclare rule. (#683)

### Bug Fixes

- Make ExportDeclaration span accurate (#928)

### Refactor

- Move Semantic into Transformer (#1130)
- Fix the lifetime annotations around Vist and VisitMut (#973)

### Testing

- Add scoping test cases (#954)

## [0.2.0] - 2023-09-14

### Features

- Implement re-exports (#877)
- Add loaded_modules to ModuleRecord
- Add runner for import-plugin (#858)
- Add `SymbolId` and `ReferenceId` (#755)
- Add `node_id` to `Reference` (#689)

### Bug Fixes

- Symbol of identifier of top level function declaration should be in the root scope (#843)
- Make semantic own `Trivias` (#711)
- Nested references (#661)

### Performance

- Lazily build trivia map instead of build in-place (#903)

### Testing

- Test harness (#679)

