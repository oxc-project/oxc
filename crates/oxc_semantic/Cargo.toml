[package]
name                   = "oxc_semantic"
version                = "0.9.0"
authors.workspace      = true
description.workspace  = true
edition.workspace      = true
homepage.workspace     = true
keywords.workspace     = true
license.workspace      = true
repository.workspace   = true
rust-version.workspace = true
categories.workspace   = true
include                = ["/src"]

[lints]
workspace = true

[lib]
doctest = false

[dependencies]
oxc_span        = { workspace = true }
oxc_ast         = { workspace = true }
oxc_syntax      = { workspace = true }
oxc_diagnostics = { workspace = true }
oxc_index       = { workspace = true }
oxc_allocator   = { workspace = true }
oxc_macros      = { workspace = true }

indexmap   = { workspace = true }
itertools  = { workspace = true }
phf        = { workspace = true, features = ["macros"] }
rustc-hash = { workspace = true }
serde      = { workspace = true, features = ["derive"], optional = true }
petgraph   = "0.6.4"

tsify        = { workspace = true, optional = true }
wasm-bindgen = { workspace = true, optional = true }

[dev-dependencies]
oxc_parser = { workspace = true }

bitflags          = { workspace = true }
indexmap          = { workspace = true }
insta             = { workspace = true }
itertools         = { workspace = true }
phf               = { workspace = true, features = ["macros"] }
pretty_assertions = "1.4.0"
rustc-hash        = { workspace = true }

[features]
default = []
serde   = ["dep:serde", "oxc_span/serde", "oxc_syntax/serde"]
wasm    = ["serde", "dep:tsify", "dep:wasm-bindgen"]
