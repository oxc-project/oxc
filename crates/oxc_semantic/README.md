# Oxc Semantic

Comprehensive semantic analysis for JavaScript and TypeScript programs.

## Overview

This crate performs semantic analysis on JavaScript and TypeScript ASTs, building symbol tables, scope trees, and control flow graphs. It provides the foundation for advanced static analysis, linting, and transformation tools.

## Key Features

- **Symbol resolution**: Build complete symbol tables with binding information
- **Scope analysis**: Construct scope trees following ECMAScript scoping rules
- **Reference tracking**: Track all variable references and their relationships
- **Control flow**: Optional control flow graph construction
- **JSDoc parsing**: Extract and parse JSDoc comments
- **Module analysis**: Analyze import/export relationships

## Architecture

### Semantic Analysis Pipeline

1. **AST Traversal**: Visit all nodes to collect declarations
2. **Scope Building**: Construct scope tree following language rules
3. **Symbol Resolution**: Create symbol table with binding information
4. **Reference Analysis**: Track all identifier references
5. **Control Flow**: Optionally build control flow graphs
6. **JSDoc Processing**: Parse and attach documentation

### Key Data Structures

#### Symbol Table

- **Symbols**: All declared identifiers (variables, functions, classes, etc.)
- **References**: All uses of identifiers
- **Bindings**: Association between names and symbols in scopes

#### Scope Tree

- **Scope hierarchy**: Nested scopes following language semantics
- **Binding resolution**: How identifiers resolve to declarations
- **Closure analysis**: Capture and usage patterns

#### Control Flow Graph

- **Basic blocks**: Sequences of statements with single entry/exit
- **Flow edges**: Conditional and unconditional control flow
- **Exception handling**: Try/catch/finally flow modeling

### Applications

- **Linting**: Detect unused variables, undefined references, etc.
- **Transformation**: Safe variable renaming and scope analysis
- **Analysis**: Dead code detection, dependency analysis
- **IDE features**: Go-to-definition, find references, refactoring

The semantic analyzer provides the deep program understanding needed for sophisticated JavaScript and TypeScript tooling.
