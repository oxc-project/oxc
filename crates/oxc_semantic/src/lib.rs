mod binder;
mod builder;
mod checker;
mod class;
mod control_flow;
mod diagnostics;
mod jsdoc;
mod label;
mod module_record;
mod node;
pub mod pg;
mod reference;
mod scope;
mod symbol;

use std::{rc::Rc, sync::Arc};

pub use petgraph;

pub use builder::{SemanticBuilder, SemanticBuilderReturn};
use class::ClassTable;
pub use jsdoc::{JSDoc, JSDocFinder, JSDocTag};
use oxc_ast::{ast::IdentifierReference, AstKind, Trivias};
use oxc_span::SourceType;
pub use oxc_syntax::{
    module_record::ModuleRecord,
    scope::{ScopeFlags, ScopeId},
    symbol::{SymbolFlags, SymbolId},
};
use rustc_hash::FxHashSet;

pub use crate::{
    control_flow::{
        print_basic_block, AssignmentValue, BasicBlockElement, BinaryAssignmentValue, BinaryOp,
        CallType, CalleeWithArgumentsAssignmentValue, CollectionAssignmentValue, ControlFlowGraph,
        EdgeType, ObjectPropertyAccessAssignmentValue, Register, UnaryExpressioneAssignmentValue,
        UpdateAssignmentValue,
    },
    node::{AstNode, AstNodeId, AstNodes},
    reference::{Reference, ReferenceFlag, ReferenceId},
    scope::ScopeTree,
    symbol::SymbolTable,
};

pub struct Semantic<'a> {
    source_text: &'a str,

    source_type: SourceType,

    nodes: AstNodes<'a>,

    scopes: ScopeTree,

    symbols: SymbolTable,

    classes: ClassTable,

    trivias: Rc<Trivias>,

    module_record: Arc<ModuleRecord>,

    jsdoc: JSDocFinder<'a>,

    unused_labels: FxHashSet<AstNodeId>,

    cfg: ControlFlowGraph,
}

impl<'a> Semantic<'a> {
    pub fn into_symbol_table_and_scope_tree(self) -> (SymbolTable, ScopeTree) {
        (self.symbols, self.scopes)
    }

    pub fn source_text(&self) -> &'a str {
        self.source_text
    }

    pub fn source_type(&self) -> &SourceType {
        &self.source_type
    }

    pub fn nodes(&self) -> &AstNodes<'a> {
        &self.nodes
    }

    pub fn scopes(&self) -> &ScopeTree {
        &self.scopes
    }

    pub fn classes(&self) -> &ClassTable {
        &self.classes
    }

    pub fn scopes_mut(&mut self) -> &mut ScopeTree {
        &mut self.scopes
    }

    pub fn trivias(&self) -> &Trivias {
        &self.trivias
    }

    pub fn jsdoc(&self) -> &JSDocFinder<'a> {
        &self.jsdoc
    }

    pub fn module_record(&self) -> &Arc<ModuleRecord> {
        &self.module_record
    }

    pub fn symbols(&self) -> &SymbolTable {
        &self.symbols
    }

    pub fn unused_labels(&self) -> &FxHashSet<AstNodeId> {
        &self.unused_labels
    }

    pub fn cfg(&self) -> &ControlFlowGraph {
        &self.cfg
    }

    pub fn is_unresolved_reference(&self, node_id: AstNodeId) -> bool {
        let reference_node = self.nodes.get_node(node_id);
        let AstKind::IdentifierReference(id) = reference_node.kind() else {
            return false;
        };
        self.scopes().root_unresolved_references().contains_key(id.name.as_str())
    }

    /// Find which scope a symbol is declared in
    pub fn symbol_scope(&self, symbol_id: SymbolId) -> ScopeId {
        self.symbols.get_scope_id(symbol_id)
    }

    /// Get all resolved references for a symbol
    pub fn symbol_references(&self, symbol_id: SymbolId) -> impl Iterator<Item = &Reference> + '_ {
        self.symbols.get_resolved_references(symbol_id)
    }

    pub fn symbol_declaration(&self, symbol_id: SymbolId) -> &AstNode<'a> {
        self.nodes.get_node(self.symbols.get_declaration(symbol_id))
    }

    pub fn is_reference_to_global_variable(&self, ident: &IdentifierReference) -> bool {
        self.scopes().root_unresolved_references().contains_key(ident.name.as_str())
    }
}

#[cfg(test)]
mod tests {
    use oxc_allocator::Allocator;
    use oxc_ast::{ast::VariableDeclarationKind, AstKind};
    use oxc_span::{Atom, SourceType};

    use super::*;

    /// Create a [`Semantic`] from source code, assuming there are no syntax/semantic errors.
    fn get_semantic<'s, 'a: 's>(
        allocator: &'a Allocator,
        source: &'s str,
        source_type: SourceType,
    ) -> Semantic<'s> {
        let parse = oxc_parser::Parser::new(allocator, source, source_type).parse();
        assert!(parse.errors.is_empty());
        let program = allocator.alloc(parse.program);
        let semantic = SemanticBuilder::new(source, source_type).build(program);
        assert!(semantic.errors.is_empty(), "Parse error: {}", semantic.errors[0]);
        semantic.semantic
    }

    #[test]
    fn test_symbols() {
        let source = "
            let a;
            function foo(a) {
                return a + 1;
            }
            let b = a + foo(1);";
        let allocator = Allocator::default();
        let semantic = get_semantic(&allocator, source, SourceType::default());

        let top_level_a = semantic
            .scopes()
            .get_binding(semantic.scopes().root_scope_id(), &Atom::from("a"))
            .unwrap();

        let decl = semantic.symbol_declaration(top_level_a);
        match decl.kind() {
            AstKind::VariableDeclarator(decl) => {
                assert_eq!(decl.kind, VariableDeclarationKind::Let);
            }
            kind => panic!("Expected VariableDeclarator for 'let', got {kind:?}"),
        }

        let references = semantic.symbol_references(top_level_a);
        assert_eq!(references.count(), 1);
    }

    #[test]
    fn test_top_level_symbols() {
        let source = "function Fn() {}";
        let allocator = Allocator::default();
        let semantic = get_semantic(&allocator, source, SourceType::default());

        let top_level_a = semantic
            .scopes()
            .iter_bindings()
            .find(|(_scope_id, _symbol_id, name)| name.as_str() == "Fn")
            .unwrap();
        assert_eq!(semantic.symbols.get_scope_id(top_level_a.1), top_level_a.0);
    }

    #[test]
    fn test_is_global() {
        let source = "
            var a = 0;
            function foo() {
            a += 1;
            }

            var b = a + 2;
        ";
        let allocator = Allocator::default();
        let semantic = get_semantic(&allocator, source, SourceType::default());
        for node in semantic.nodes().iter() {
            if let AstKind::IdentifierReference(id) = node.kind() {
                assert!(!semantic.is_reference_to_global_variable(id));
            }
        }
    }

    #[test]
    fn type_alias_gets_reference() {
        let source = "type A = 1; type B = A";
        let allocator = Allocator::default();
        let source_type: SourceType = SourceType::default().with_typescript(true);
        let semantic = get_semantic(&allocator, source, source_type);
        assert!(semantic.symbols().references.len() == 1);
    }

    #[test]
    fn test_reference_resolutions_simple_read_write() {
        let alloc = Allocator::default();
        let target_symbol_name = Atom::from("a");
        let typescript = SourceType::default().with_typescript(true).with_module(true);
        let sources = [
            // simple cases
            (SourceType::default(), "let a = 1; a = 2", ReferenceFlag::write()),
            (SourceType::default(), "let a = 1, b; b = a", ReferenceFlag::read()),
            (SourceType::default(), "let a = 1, b; b[a]", ReferenceFlag::read()),
            (SourceType::default(), "let a = 1, b = 1, c; c = a + b", ReferenceFlag::read()),
            (SourceType::default(), "function a() { return }; a()", ReferenceFlag::read()),
            (SourceType::default(), "class a {}; new a()", ReferenceFlag::read()),
            (SourceType::default(), "let a; function foo() { return a }", ReferenceFlag::read()),
            // pattern assignment
            (SourceType::default(), "let a = 1, b; b = { a }", ReferenceFlag::read()),
            (SourceType::default(), "let a, b; ({ b } = { a })", ReferenceFlag::read()),
            (SourceType::default(), "let a, b; ({ a } = { b })", ReferenceFlag::write()),
            (SourceType::default(), "let a, b; ([ b ] = [ a ])", ReferenceFlag::read()),
            (SourceType::default(), "let a, b; ([ a ] = [ b ])", ReferenceFlag::write()),
            // property access/mutation
            (SourceType::default(), "let a = { b: 1 }; a.b = 2", ReferenceFlag::read()),
            (SourceType::default(), "let a = { b: 1 }; a.b += 2", ReferenceFlag::read()),
            // parens are pass-through
            (SourceType::default(), "let a = 1, b; b = (a)", ReferenceFlag::read()),
            (SourceType::default(), "let a = 1, b; b = ++(a)", ReferenceFlag::read_write()),
            (SourceType::default(), "let a = 1, b; b = ++((((a))))", ReferenceFlag::read_write()),
            (SourceType::default(), "let a = 1, b; b = ((++((a))))", ReferenceFlag::read_write()),
            // simple binops/calls for sanity check
            (SourceType::default(), "let a, b; a + b", ReferenceFlag::read()),
            (SourceType::default(), "let a, b; b(a)", ReferenceFlag::read()),
            (SourceType::default(), "let a, b; a = 5", ReferenceFlag::write()),
            // unary op counts as write, but checking continues up tree
            (SourceType::default(), "let a = 1, b; b = ++a", ReferenceFlag::read_write()),
            (SourceType::default(), "let a = 1, b; b = --a", ReferenceFlag::read_write()),
            (SourceType::default(), "let a = 1, b; b = a++", ReferenceFlag::read_write()),
            (SourceType::default(), "let a = 1, b; b = a--", ReferenceFlag::read_write()),
            // assignment expressions count as read-write
            (SourceType::default(), "let a = 1, b; b = a += 5", ReferenceFlag::read_write()),
            (SourceType::default(), "let a = 1; a += 5", ReferenceFlag::read_write()),
            // note: we consider a to be written, and the read of `1` propagates upwards
            (SourceType::default(), "let a, b; b = a = 1", ReferenceFlag::read_write()),
            (SourceType::default(), "let a, b; b = (a = 1)", ReferenceFlag::read_write()),
            (SourceType::default(), "let a, b, c; b = c = a", ReferenceFlag::read()),
            // sequences return last value in sequence
            (SourceType::default(), "let a, b; b = (0, a++)", ReferenceFlag::read_write()),
            // loops
            (
                SourceType::default(),
                "var a, arr = [1, 2, 3]; for(a in arr) { break }",
                ReferenceFlag::write(),
            ),
            (
                SourceType::default(),
                "var a, obj = { }; for(a of obj) { break }",
                ReferenceFlag::write(),
            ),
            (SourceType::default(), "var a; for(; false; a++) { }", ReferenceFlag::read_write()),
            (SourceType::default(), "var a = 1; while(a < 5) { break }", ReferenceFlag::read()),
            // if statements
            (SourceType::default(), "let a; if (a) { true } else { false }", ReferenceFlag::read()),
            (
                SourceType::default(),
                "let a, b; if (a == b) { true } else { false }",
                ReferenceFlag::read(),
            ),
            (
                SourceType::default(),
                "let a, b; if (b == a) { true } else { false }",
                ReferenceFlag::read(),
            ),
            // identifiers not in last value are also considered a read (at
            // least, or now)
            (SourceType::default(), "let a, b; b = (a, 0)", ReferenceFlag::read()),
            (SourceType::default(), "let a, b; b = (--a, 0)", ReferenceFlag::read_write()),
            // other reads after a is written
            // a = 1 writes, but the CallExpression reads the rhs (1) so a isn't read
            (
                SourceType::default(),
                "let a; function foo(a) { return a }; foo(a = 1)",
                ReferenceFlag::read_write(),
            ),
            // member expression
            (SourceType::default(), "let a; a.b = 1", ReferenceFlag::read()),
            (SourceType::default(), "let a; let b; b[a += 1] = 1", ReferenceFlag::read_write()),
            (
                SourceType::default(),
                "let a; let b; let c; b[c[a = c['a']] = 'c'] = 'b'",
                ReferenceFlag::read_write(),
            ),
            (
                SourceType::default(),
                "let a; let b; let c; a[c[b = c['a']] = 'c'] = 'b'",
                ReferenceFlag::read(),
            ),
            // typescript
            (typescript, "let a: number = 1; (a as any) = true", ReferenceFlag::write()),
            (typescript, "let a: number = 1; a = true as any", ReferenceFlag::write()),
            (typescript, "let a: number = 1; a = 2 as const", ReferenceFlag::write()),
            (typescript, "let a: number = 1; a = 2 satisfies number", ReferenceFlag::write()),
            (typescript, "let a: number; (a as any) = 1;", ReferenceFlag::write()),
        ];

        for (source_type, source, flag) in sources {
            let semantic = get_semantic(&alloc, source, source_type);
            let a_id =
                semantic.scopes().get_root_binding(&target_symbol_name).unwrap_or_else(|| {
                    panic!("no references for '{target_symbol_name}' found");
                });
            let a_refs: Vec<_> = semantic.symbol_references(a_id).collect();
            let num_refs = a_refs.len();

            assert!(num_refs == 1, "expected to find 1 reference to '{target_symbol_name}' but {num_refs} were found\n\nsource:\n{source}");
            let ref_type = a_refs[0];
            if flag.is_write() {
                assert!(
                    ref_type.is_write(),
                    "expected reference to '{target_symbol_name}' to be write\n\nsource:\n{source}"
                );
            } else {
                assert!(!ref_type.is_write(), "expected reference to '{target_symbol_name}' not to have been written to, but it is\n\nsource:\n{source}");
            }
            if flag.is_read() {
                assert!(
                    ref_type.is_read(),
                    "expected reference to '{target_symbol_name}' to be read\n\nsource:\n{source}"
                );
            } else {
                assert!(!ref_type.is_read(), "expected reference to '{target_symbol_name}' not to be read, but it is\n\nsource:\n{source}");
            }
        }
    }
}
