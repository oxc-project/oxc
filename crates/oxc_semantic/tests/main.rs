use std::{fs, path::Path};

use oxc_allocator::Allocator;
use oxc_index::Idx;
use oxc_parser::Parser;
use oxc_semantic::{ScopeId, Semantic, SemanticBuilder};
use oxc_span::SourceType;

fn get_scope_snapshot(semantic: &Semantic, scopes: impl Iterator<Item = ScopeId>) -> String {
    let scope_tree = semantic.scopes();
    let mut result = String::default();

    result.push('[');
    scopes.enumerate().for_each(|(index, scope_id)| {
        if index != 0 {
            result.push(',');
        }
        let flags = scope_tree.get_flags(scope_id);
        result.push('{');
        let child_ids = semantic
            .scopes()
            .descendants_from_root()
            .filter(|id| {
                scope_tree.get_parent_id(*id).is_some_and(|parent_id| parent_id == scope_id)
            })
            .collect::<Vec<_>>();
        result.push_str("\"children\":");
        result.push_str(&get_scope_snapshot(semantic, child_ids.iter().copied()));
        result.push(',');
        result.push_str(format!("\"flags\": \"{flags:?}\",").as_str());
        result.push_str(format!("\"id\": {},", scope_id.index()).as_str());
        result.push_str(
            format!(
                "\"node\": {:?},",
                semantic.nodes().kind(scope_tree.get_node_id(scope_id)).debug_name()
            )
            .as_str(),
        );
        result.push_str("\"symbols\": ");
        let bindings = scope_tree.get_bindings(scope_id);
        result.push('[');
        bindings.iter().enumerate().for_each(|(index, (name, symbol_id))| {
            if index != 0 {
                result.push(',');
            }
            result.push('{');
            result.push_str(
                format!("\"flags\": \"{:?}\",", semantic.symbols().get_flags(*symbol_id)).as_str(),
            );
            result.push_str(format!("\"id\": {},", symbol_id.index()).as_str());
            result.push_str(format!("\"name\": {name:?},").as_str());
            result.push_str(
                format!(
                    "\"node\": {:?},",
                    semantic
                        .nodes()
                        .kind(semantic.symbols().get_declaration(*symbol_id))
                        .debug_name()
                )
                .as_str(),
            );
            {
                result.push_str("\"references\": ");
                result.push('[');
                semantic
                    .symbols()
                    .get_resolved_reference_ids(*symbol_id)
                    .iter()
                    .enumerate()
                    .for_each(|(index, reference_id)| {
                        if index != 0 {
                            result.push(',');
                        }
                        let reference = &semantic.symbols().references[*reference_id];
                        result.push('{');
                        result
                            .push_str(format!("\"flags\": \"{:?}\",", reference.flags()).as_str());
                        result.push_str(format!("\"id\": {},", reference_id.index()).as_str());
                        result.push_str(
                            format!("\"name\": {:?},", semantic.reference_name(reference)).as_str(),
                        );
                        result.push_str(
                            format!("\"node_id\": {}", reference.node_id().index()).as_str(),
                        );
                        result.push('}');
                    });
                result.push(']');
            }
            result.push('}');
        });
        result.push(']');
        result.push('}');
    });
    result.push(']');
    result
}

fn analyze(path: &Path, source_text: &str) -> String {
    let allocator = Allocator::default();
    let source_type = SourceType::from_path(path).unwrap();

    let ret = Parser::new(&allocator, source_text, source_type).parse();
    let semantic = SemanticBuilder::new(source_text).build(&ret.program).semantic;

    let scopes = get_scope_snapshot(&semantic, vec![semantic.scopes().root_scope_id()].into_iter());
    let value: serde_json::Value = serde_json::from_str(scopes.as_str()).unwrap();
    serde_json::to_string_pretty(&value).unwrap()
}

/// # Panics
/// cargo test --package oxc_semantic --test main
#[test]
fn main() {
    insta::glob!("fixtures/**/*.{js,jsx,ts,tsx}", |path| {
        let source_text = fs::read_to_string(path).unwrap();
        let snapshot = analyze(path, &source_text);
        let name = path.file_stem().unwrap().to_str().unwrap();
        insta::with_settings!({ snapshot_path => path.parent().unwrap(), prepend_module_to_snapshot => false, snapshot_suffix => "", omit_expression => true }, {
            insta::assert_snapshot!(name, snapshot);
        });
    });
}
