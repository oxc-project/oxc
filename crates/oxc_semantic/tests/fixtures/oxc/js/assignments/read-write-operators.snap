---
source: crates/oxc_semantic/tests/main.rs
input_file: crates/oxc_semantic/tests/fixtures/oxc/js/assignments/read-write-operators.js
---
[
  {
    "children": [],
    "flags": "ScopeFlags(StrictMode | Top)",
    "id": 0,
    "node": "Program",
    "symbols": [
      {
        "flags": "SymbolFlags(BlockScopedVariable)",
        "id": 0,
        "name": "Foo",
        "node": "VariableDeclarator(Foo)",
        "references": [
          {
            "flags": "ReferenceFlags(Write)",
            "id": 0,
            "name": "Foo",
            "node_id": 9
          },
          {
            "flags": "ReferenceFlags(Read | Write)",
            "id": 1,
            "name": "Foo",
            "node_id": 15
          },
          {
            "flags": "ReferenceFlags(Read | Write)",
            "id": 2,
            "name": "Foo",
            "node_id": 21
          },
          {
            "flags": "ReferenceFlags(Read | Write)",
            "id": 3,
            "name": "Foo",
            "node_id": 27
          },
          {
            "flags": "ReferenceFlags(Read | Write)",
            "id": 4,
            "name": "Foo",
            "node_id": 33
          },
          {
            "flags": "ReferenceFlags(Read | Write)",
            "id": 5,
            "name": "Foo",
            "node_id": 39
          },
          {
            "flags": "ReferenceFlags(Read | Write)",
            "id": 6,
            "name": "Foo",
            "node_id": 45
          },
          {
            "flags": "ReferenceFlags(Read | Write)",
            "id": 7,
            "name": "Foo",
            "node_id": 51
          },
          {
            "flags": "ReferenceFlags(Read | Write)",
            "id": 8,
            "name": "Foo",
            "node_id": 57
          },
          {
            "flags": "ReferenceFlags(Read | Write)",
            "id": 9,
            "name": "Foo",
            "node_id": 63
          },
          {
            "flags": "ReferenceFlags(Read | Write)",
            "id": 10,
            "name": "Foo",
            "node_id": 69
          },
          {
            "flags": "ReferenceFlags(Read | Write)",
            "id": 11,
            "name": "Foo",
            "node_id": 75
          },
          {
            "flags": "ReferenceFlags(Read | Write)",
            "id": 12,
            "name": "Foo",
            "node_id": 81
          },
          {
            "flags": "ReferenceFlags(Read | Write)",
            "id": 13,
            "name": "Foo",
            "node_id": 87
          },
          {
            "flags": "ReferenceFlags(Read | Write)",
            "id": 14,
            "name": "Foo",
            "node_id": 93
          },
          {
            "flags": "ReferenceFlags(Read | Write)",
            "id": 15,
            "name": "Foo",
            "node_id": 99
          }
        ]
      }
    ]
  }
]
